src/treasury.cairo

```diff
@@ -190,6 +232,57 @@ mod Treasury {
             return status;
         }
 
+        // TODO: Add transfers querying methods for frontend
+        
+        fn add_transfer(
+            ref self: ContractState,
+            receiver: ContractAddress,
+            amount: u256,
+            token_addr: ContractAddress
+        ) {
+            let token: IERC20Dispatcher = IERC20Dispatcher { contract_address: token_addr };
+            assert(token.balanceOf(get_contract_address()) >= amount, Errors::INSUFFICIENT_FUNDS);
+            let new_transfer_id = self.last_transfer_id.read() + 1;
+            self.last_transfer_id.write(new_transfer_id);
+            let transfer = Transfer {receiver, token_addr, amount, is_finished: false, is_cancelled: false};
+            self.transfers_on_cooldown.write(new_transfer_id, transfer);
+            self.emit(TransferPending { receiver, token_addr, amount });
+        }
+
+        fn cancel_transaction(ref self: ContractState, transfer_id: u64) {
+            // assert if guardian
+            let initial_transfer = self.transfers_on_cooldown.read(transfer_id);
+            let cancelation_event = TransferCancelled { 
+                initial_amount: initial_transfer.amount,
+                token_addr: initial_transfer.token_addr,
+                receiver: initial_transfer.receiver 
+            };
+            self.transfers_on_cooldown.write(transfer_id, Transfer {amount: 0, is_cancelled: true, ..initial_transfer});
+            self.emit(cancelation_event);
+        }
+
+        // Who can execute it?
+        fn execute_current_pending(ref self: ContractState) -> bool {
+            let current_id = self.current_transfer_pointer.read();
+            let transfer_pending = self.transfers_on_cooldown.read(current_id);
+            self.current_transfer_pointer.write(current_id + 1);
+            if transfer_pending.is_cancelled {
+                return false; // TODO: Add failure event?
+            }
+            let token: IERC20Dispatcher = IERC20Dispatcher { contract_address: transfer_pending.token_addr };
+            let status: bool = token.transfer(transfer_pending.receiver, transfer_pending.amount);
```

Comment:
If transaction executed but was reverted after that, it will be marked finished. What behavior should we expect on transaction revert, some other event? Or boolean value to mark it as reverted for frontend?

Reply:
That is not how transactions work. If a tx reverts any time, all of its effects are nil – it's as if it never were executed.